Feature/0002 bugfix and improvements

ajax-torbin-v opened the pull request.

ajax-torbin-v committed:
Added detekt plugin. Bit of refactoring

ajax-torbin-v committed:
Added bulk operations for product quantity update.
Fixed & added update functions.
Added corresponding DTO.

ajax-butriy-v commented on src/main/kotlin/com/example/delivery/controller/UserController.kt:
```diff
@@ -28,6 +30,11 @@ class UserController(private val userService: UserService) {
         return userService.getById(id).toDTO()
     }
 
+    @PatchMapping("/id")
```
"**/{id}**"

ajax-makarytskyi-o commented on src/main/kotlin/com/example/delivery/service/UserService.kt:
```diff
@@ -26,6 +29,20 @@ class UserService(
+        return update
+    }
+
     fun findAllOrders(id: String): List<DomainOrder> {
```
findAllOrders should be renamed to getAllOrders, because this method throws exception.

ajax-torbin-v committed:
Small fixes

ajax-shulepa-a commented on src/main/kotlin/com/example/delivery/service/OrderService.kt:
```diff
@@ -35,27 +37,37 @@ class OrderService(
+
+    private fun createUpdateObject(updateOrderDTO: UpdateOrderDTO): Update {
+        val update = Update()
+        with(updateOrderDTO) {
+            status?.let { update.set(MongoOrder::status.name, MongoOrder.Status.valueOf(status)) }
+            shipmentDetails?.let { update.set(MongoOrder::shipmentDetails.name, shipmentDetails) }
```
service is responsible for building mongo updates. looks like something wrong with SRP here

ajax-shulepa-a commented on src/main/kotlin/com/example/delivery/service/ProductService.kt:
```diff
@@ -27,4 +30,19 @@ class ProductService(private val productRepository: ProductRepository) {
+    private fun createUpdateObject(updateProductDTO: UpdateProductDTO): Update {
+        val update = Update()
+        with(updateProductDTO) {
+            name?.let { update.set(MongoProduct::name.name, this) }
+            price?.let { update.set(MongoProduct::price.name, this) }
+            amountAvailable?.let { update.set(MongoProduct::amountAvailable.name, this) }
```
what is this here? where are tests to check that this code works correctly?

and same comment about SRP

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/controller/UserController.kt:
```diff
@@ -28,6 +30,11 @@ class UserController(private val userService: UserService) {
         return userService.getById(id).toDTO()
     }
 
+    @PatchMapping("/{id}")
```
This is multipurpose API which I see is used to change Order lifecycle and Order props like ship details. I think better to have separate API for pure update details and change lifecycle. Then you will not need to have such Patch at all since PUT is more suitable for update. Patch is more complex to maintain

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/service/UserService.kt:
```diff
@@ -26,7 +29,21 @@ class UserService(
+    private fun createUpdateObject(updateUserDTO: UpdateUserDTO): Update {
+        val update = Update()
+        with(updateUserDTO) {
+            fullName?.let { update.set(MongoUser::fullName.name, fullName) }
```
same what Andrii said

ajax-torbin-v committed:
Changed patch methods into put in controllers
Added separate method for patching order's status
Moved creating Update object into mappers
Added tests for update methods for controller and service layers
Created corresponding fixture objects
